### PR TITLE
Add catalyst to the documentation

### PR DESCRIPTION
**Pull request by josh146**
**Context:** Adds Catalyst to the PL documentation and readme.

**Possible Drawbacks:** We may need to update the examples to keep in sync with Catalyst.

**Related GitHub Issues:** n/a

### CODE REVIEW

File: doc/introduction/compiling_circuits.rst [hunk 1]
```diff
@@ -332,3 +338,69 @@ observables and coefficients:
...
+
+    import jaxopt
+
+    @qjit
+    def optimization():
+        # initial parameter
+        params = jnp.array([0.54, 0.3154])
+
+        # define the optimizer
+        opt = jaxopt.GradientDescent(cost, stepsize=0.4)
```
Comment (josh146) [inline doc/introduction/compiling_circuits.rst:390]: @dime10 my example here is assuming the JAX integration, which actually really simplifies this example -- you don't need to wrap the cost function to manually compute the gradient!

Comment (dime10) [reply doc/introduction/compiling_circuits.rst:390]: Does this work given that it's inside the qjit decorator?

I think it would only work if we had a qnode primitive inside the JAXPR.

Comment (josh146) [reply doc/introduction/compiling_circuits.rst:390]: @dime10 you're right, I got ahead of myself

Comment (josh146) [reply doc/introduction/compiling_circuits.rst:390]: I think it would work if we did

```python
@jax.jit
@qjit
@qnode(dev)
```
lol

Comment (dime10) [reply doc/introduction/compiling_circuits.rst:390]: The `jax.jit` on the qnode is not necessary, but I suppose you could do :
```py
@qjit
@qml.qnode
def cost():
   ...


@jax.jit
def workflow():
   ...
   jaxopt.GradientDescent(cost, stepsize=0.4)
```

Comment (josh146) [reply doc/introduction/compiling_circuits.rst:390]: Yep, this is exactly what I was going for 🤦 Thanks @dime10

File: doc/introduction/compiling_circuits.rst [hunk 2]
```diff
@@ -332,3 +338,69 @@ observables and coefficients:
...
+
+The ``qjit`` decorator can also be used on hybrid cost functions -- that is,
+cost functions that include both QNodes and classical processing. We can even
+JIT compile the full optimization loop, for example when training models:
+
+.. code-block:: python
+
+    import jaxopt
+
+    @qjit
```
Comment (josh146) [inline doc/introduction/compiling_circuits.rst:384]:
```suggestion
    @jax.jit
```

Comment (josh146) [reply doc/introduction/compiling_circuits.rst:384]: @dime10 this would fix the example, with the downside that it is not yet ready in Catalyst

Comment (josh146) [reply doc/introduction/compiling_circuits.rst:384]: However, the PL documentation can only be updated with a PL release, so if we fallback to the old/current way of doing it, it will be outdated soon

File: doc/index.rst [hunk 3]
```diff
@@ -71,6 +71,12 @@ Features
 
 ..
 
+- *Just in time compilation*. Experimental support for just-in-time
+  compilation. Compile your entire hybrid workflow, loops. See
+  `Catalyst <https://github.com/pennylaneai/catalyst>`__ for more details.
+
+..
```
[the remark this answers is not in the document]
Comment (trbromley) [reply doc/index.rst:78]: Second minor point - with the added content, the last bullet point ends up being full width and doesn't look as great. Any idea if there's an easy fix for this?
image

File: doc/introduction/compiling_circuits.rst [hunk 4]
```diff
@@ -15,6 +15,12 @@ or replace a large circuit by a number of smaller circuits.
 Compilation functionality is mostly designed as **transforms**, which you can read up on in the
 section on :doc:`inspecting circuits </introduction/inspecting_circuits>`.
 
+PennyLane also supports experimental just-in-time compilation, via
+[Catalyst](https://github.com/pennylaneai/catalyst). This is more general, and
```
Comment (trbromley) [inline doc/introduction/compiling_circuits.rst:19]:
```suggestion
`Catalyst <https://github.com/pennylaneai/catalyst>`_. This is more general, and
```

File: doc/introduction/compiling_circuits.rst [hunk 5]
```diff
@@ -15,6 +15,12 @@ or replace a large circuit by a number of smaller circuits.
 Compilation functionality is mostly designed as **transforms**, which you can read up on in the
 section on :doc:`inspecting circuits </introduction/inspecting_circuits>`.
 
+PennyLane also supports experimental just-in-time compilation, via
```
Comment (trbromley) [inline doc/introduction/compiling_circuits.rst:18]: > In addition to quantum circuit transformations

This phrase from below might help smooth the transition between the two compilation concepts.

**Review by trbromley — COMMENTED** [review]
Thanks @josh146!

File: README.md [hunk 6]
```diff
@@ -47,6 +47,10 @@
 
 - *Machine learning on quantum hardware*. Connect to quantum hardware using **PyTorch**, **TensorFlow**, **JAX**, **Keras**, or **NumPy**. Build rich and flexible hybrid quantum-classical models.
 
+- *Just in time compilation*. Experimental support for just-in-time
+  compilation. Compile your entire hybrid workflow, loops. See
```
[the remark this answers is not in the document]
Comment (josh146) [reply README.md:51]: oops, no I just missed a word!

File: README.md [hunk 7]
```diff
@@ -47,6 +47,10 @@
 
 - *Machine learning on quantum hardware*. Connect to quantum hardware using **PyTorch**, **TensorFlow**, **JAX**, **Keras**, or **NumPy**. Build rich and flexible hybrid quantum-classical models.
 
+- *Just in time compilation*. Experimental support for just-in-time
+  compilation. Compile your entire hybrid workflow, loops. See
+  [Catalyst](https://github.com/pennylaneai/catalyst) for more details.
```
[the remark this answers is not in the document]
Comment (josh146) [reply README.md:52]: I think so yes

File: doc/index.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (josh146) [reply doc/index.rst:78]:
```suggestion
- *Just in time compilation*. Compile your **entire hybrid workflow**, including loops. See
  `Catalyst <https://github.com/pennylaneai/catalyst>`__ for more details.

..
```

File: doc/introduction/compiling_circuits.rst (hunk 4 above)
Comment (josh146) [reply doc/introduction/compiling_circuits.rst:19]: 🤦

File: doc/introduction/compiling_circuits.rst (hunk 5 above)
Comment (josh146) [reply doc/introduction/compiling_circuits.rst:18]:
```suggestion
In addition to quantum circuit transforms, PennyLane also
supports experimental just-in-time compilation, via
```

File: doc/introduction/compiling_circuits.rst [hunk 8]
```diff
@@ -332,3 +339,69 @@ observables and coefficients:
...
+JIT compile the full optimization loop, for example when training models:
+
+.. code-block:: python
+
+    import jaxopt
+
+    @jax.jit
+    def optimization():
+        # initial parameter
+        params = jnp.array([0.54, 0.3154])
```
Comment (dime10) [inline doc/introduction/compiling_circuits.rst:388]: I'm a bit torn here with this example because it's talking about the qjit decorator compiling hybrid cost functions but then we use JAX to do it in the example, which is supported with the new JAX integration and convenient, but still a bit strange 🤔

**Review by dime10 — COMMENTED** [review]
Thanks @josh146 for drafting up this page! :)

I have a couple of comments:

File: README.md [hunk 9]
```diff
@@ -47,6 +47,10 @@
 
 - *Machine learning on quantum hardware*. Connect to quantum hardware using **PyTorch**, **TensorFlow**, **JAX**, **Keras**, or **NumPy**. Build rich and flexible hybrid quantum-classical models.
 
+- *Just in time compilation*. Experimental support for just-in-time
+  compilation. Compile your entire hybrid workflow, including loops. See
```
[the remark this answers is not in the document]
Comment (josh146) [reply README.md:51]: Good call! 
```suggestion
  compilation. Compile your entire hybrid workflow, with support for 
  advanced features such as adaptive circuits, real-time measurement 
  feedback, and unbounded loops. See
```

File: doc/introduction/compiling_circuits.rst [hunk 10]
```diff
@@ -15,6 +15,13 @@ or replace a large circuit by a number of smaller circuits.
 Compilation functionality is mostly designed as **transforms**, which you can read up on in the
 section on :doc:`inspecting circuits </introduction/inspecting_circuits>`.
 
+In addition to quantum circuit transforms, PennyLane also
+supports experimental just-in-time compilation, via
+`Catalyst <https://github.com/pennylaneai/catalyst>`__. This is more general, and
+supports full hybrid compilation --- compiling both the classical and quantum components
+of your workflow into a binary that can be run close to the accelerators (GPUs, QPUs, CPUs)
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/compiling_circuits.rst:22]:
```suggestion
of your workflow into a binary that can be run close to the accelerators.
```

File: doc/introduction/compiling_circuits.rst [hunk 11]
```diff
@@ -332,3 +339,69 @@ observables and coefficients:
...
+
+    dev = qml.device("lightning.qubit", wires=2, shots=1000)
+
+    @qml.qnode(dev)
+    def cost(params):
+        qml.Hadamard(0)
+        qml.RX(jnp.sin(params[0]) ** 2, wires=1)
+        qml.CRY(params[0], wires=[0, 1])
+        qml.RX(jnp.sqrt(params[1]), wires=1)
+        return qml.expval(qml.PauliZ(1))
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/compiling_circuits.rst:375]: 🤦

File: doc/introduction/compiling_circuits.rst [hunk 12]
```diff
@@ -332,3 +339,69 @@ observables and coefficients:
...
+compile your hybrid workflows:
+
+.. code-block:: python
+
+    from catalyst import qjit
+    from jax import numpy as jnp
+
+    dev = qml.device("lightning.qubit", wires=2, shots=1000)
+
+    @qml.qnode(dev)
```
Comment (josh146) [inline doc/introduction/compiling_circuits.rst:369]:
```suggestion

    @qjit
    @qml.qnode(dev)
```

File: doc/introduction/compiling_circuits.rst (hunk 8 above)
Comment (josh146) [reply doc/introduction/compiling_circuits.rst:388]: Yep I'm torn as well. The one advantage is that you avoid having to create a wrapper around the cost function which returns cost _and_ gradient, and then making use of the equivalent jaxopt feature 🤔

File: doc/introduction/compiling_circuits.rst [hunk 13]
```diff
@@ -332,3 +339,69 @@ observables and coefficients:
...
+
+Just-in-time compilation with Catlyst
+-------------------------------------
+
+In addition to quantum circuit transformations, PennyLane also supports full
+hybrid just-in-time (JIT) compilation via `Catalyst
+<https://github.com/pennylaneai/catalyst>`__. Catalyst allows you to compile
+the entire quantum-classical workflow, including any optimization loops,
+which allows for optimized performance, and the ability to run the entire
+workflow on accelerator devices as appropriate.
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/compiling_circuits.rst:352]: @dime10 would you like to make an edit here for that? :)

**Review by trbromley — APPROVED** [review]

File: doc/index.rst [hunk 14]
```diff
@@ -71,6 +71,11 @@ Features
 
 ..
 
+- *Just in time compilation*. Compile your **entire hybrid workflow**, including loops. See
```
Comment (dime10) [inline doc/index.rst:74]: @trbromley should we make the same edit here as in the readme, or would that be too much text?
```suggestion
- *Just in time compilation*. Compile your **entire hybrid workflow**, with support for
  adaptive circuits, real-time measurement feedback, unbounded loops, and more. See
```

File: doc/introduction/compiling_circuits.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (dime10) [reply doc/introduction/compiling_circuits.rst:352]: We can revisit at a later date!

File: doc/introduction/compiling_circuits.rst (hunk 8 above)
Comment (dime10) [reply doc/introduction/compiling_circuits.rst:389]: Sounds good, we'll leave it like this for now then.

**Review by dime10 — APPROVED** [review]
I had a look at the docs locally, looks great!
